File: python/2326.spiral-matrix-iv.py
class Solution:
    def spiralMatrix(self, m: int, n: int, head: ListNode | None) -> list[list[int]]:
        def linked_list_gen(node: ListNode | None) -> Iterator[int]:

            while node:
                yield node.val
                node = node.next

            while True:
                yield -1

        left, right, top, bottom = 0, n - 1, 0, m - 1
        nodes = linked_list_gen(head)
        res = [[-1] * n for _ in range(m)]

        while top < bottom and left < right:
            for i in range(left, right):
                res[top][i] = next(nodes)

            for i in range(top, bottom):
                res[i][right] = next(nodes)

            for i in range(right, left, -1):
                res[bottom][i] = next(nodes)

            for i in range(bottom, top, -1):
                res[i][left] = next(nodes)

            left += 1
            right -= 1
            top += 1
            bottom -= 1

        for i in range(left, right + 1):
            for j in range(top, bottom + 1):
                res[j][i] = next(nodes)

        return res

File: cpp/2326.spiral-matrix-iv.cpp
class Solution {

public:
  vector<vector<int>> spiralMatrix(int m, int n, ListNode *head) {

    auto getNext = [&head]() -> int {
      if (head) {
        int x = head->val;
        head = head->next;
        return x;
      }
      return -1;
    };

    vector<vector<int>> res(m, vector<int>(n, -1));

    int top = 0, left = 0, bottom = m - 1, right = n - 1;

    while (top < bottom && left < right) {

      for (int i = left; i < right; i++)
        res[top][i] = getNext();

      for (int i = top; i < bottom; i++)
        res[i][right] = getNext();

      for (int i = right; i > left; i--)
        res[bottom][i] = getNext();

      for (int i = bottom; i > top; i--)
        res[i][left] = getNext();

      top++, left++, bottom--, right--;
    }

    for (int i = top; i <= bottom; i++)
      for (int j = left; j <= right; j++)
        res[i][j] = getNext();

    return res;
  }
};